### ISSUE TITLE
seg fault in argo-rollouts pod

### ISSUE THREAD

**Issue by amit-handda** [argoproj/argo-rollouts#311]
I am using minikube(1.4.0) and k8s(1.16.0) on mac(catalina). 
Installed argo-rollouts in the above cluster, as described in the instructions on the site. 
Created a rollout of nginx to test the blue-green strategy (attached).

Faced **controller seg-fault** when I update the rollout template to upgrade/downgrade nginx image version (1.15 -> 1.16 -> 1.17, eg). It applied 1.15->1.16 correctly but crashed while applying the second transition (1.16->1.17). It crashed similarly in another iteration while I was doing second transition.

I faced it twice, consistently . Hence, reporting it. is it a known issue ? 

```
time="2019-11-27T22:58:05Z" level=info msg="Creating event broadcaster"
time="2019-11-27T22:58:05Z" level=info msg="Setting up event handlers"
time="2019-11-27T22:58:05Z" level=info msg="Setting up experiments event handlers"
time="2019-11-27T22:58:05Z" level=info msg="Setting up analysis event handlers"
time="2019-11-27T22:58:05Z" level=info msg="Waiting for controller's informer caches to sync"
time="2019-11-27T22:58:05Z" level=info msg="Enqueuing parent of default/nginx-66ffbc8db7: Rollout default/nginx"
time="2019-11-27T22:58:05Z" level=info msg="Enqueuing parent of default/nginx-6d464c748b: Rollout default/nginx"
time="2019-11-27T22:58:05Z" level=info msg="Starting Controllers"
time="2019-11-27T22:58:05Z" level=info msg="Started controller"
time="2019-11-27T22:58:05Z" level=info msg="Starting Metric Server at 0.0.0.0:8090"
time="2019-11-27T22:58:05Z" level=info msg="Starting Service workers"
time="2019-11-27T22:58:05Z" level=info msg="Started Service workers"
time="2019-11-27T22:58:05Z" level=info msg="Successfully synced" namespace=kube-system service=kube-dns
time="2019-11-27T22:58:05Z" level=info msg="Successfully synced" namespace=default service=argo-rollouts-metrics
time="2019-11-27T22:58:05Z" level=info msg="Starting Experiment workers"
time="2019-11-27T22:58:05Z" level=info msg="Started Experiment workers"
time="2019-11-27T22:58:05Z" level=info msg="Successfully synced" namespace=default service=kubernetes
time="2019-11-27T22:58:05Z" level=info msg="Successfully synced" namespace=default service=nginx-1
time="2019-11-27T22:58:05Z" level=info msg="Successfully synced" namespace=default service=nginx-2
time="2019-11-27T22:58:05Z" level=info msg="Starting Rollout workers"
time="2019-11-27T22:58:05Z" level=info msg="Started Rollout workers"
time="2019-11-27T22:58:05Z" level=info msg="Started syncing rollout at (2019-11-27 22:58:05.282240502 +0000 UTC m=+0.141368893)" namespace=default rollout=nginx
time="2019-11-27T22:58:05Z" level=info msg="Starting analysis workers"
time="2019-11-27T22:58:05Z" level=info msg="Started 30 analysis workers"
time="2019-11-27T22:58:05Z" level=info msg="Syncing replicas only (controllerPaused: true, userPaused false, isScaling: false)" namespace=default rollout=nginx
time="2019-11-27T22:58:05Z" level=info msg="Reconciliation completed" namespace=default rollout=nginx time_ms=3.262665
E1127 22:58:05.285620       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 197 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1641b00, 0x299e360)
        /go/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1641b00, 0x299e360)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/argoproj/argo-rollouts/rollout.(*RolloutController).reconcileBlueGreenPause(0xc0000cc300, 0xc0005ff5e8, 0xc0005ff3f0, 0xc000842f00, 0x2)
        /go/src/github.com/argoproj/argo-rollouts/rollout/bluegreen.go:147 +0x87
github.com/argoproj/argo-rollouts/rollout.(*RolloutController).syncReplicasOnly(0xc0000cc300, 0xc0003d0000, 0xc0001af540, 0x2, 0x2, 0x0, 0x0, 0x0)
        /go/src/github.com/argoproj/argo-rollouts/rollout/sync.go:228 +0x678
github.com/argoproj/argo-rollouts/rollout.(*RolloutController).syncHandler(0xc0000cc300, 0xc0006445f0, 0xd, 0x0, 0x0)
        /go/src/github.com/argoproj/argo-rollouts/rollout/controller.go:286 +0x9c9
github.com/argoproj/argo-rollouts/utils/controller.processNextWorkItem.func1(0x1ae6b60, 0xc00000dbc0, 0xc00058de88, 0xc0004c4600, 0x183247b, 0x7, 0x15ce2e0, 0xc00081bd50, 0x0, 0x0)
        /go/src/github.com/argoproj/argo-rollouts/utils/controller/controller.go:65 +0x131
github.com/argoproj/argo-rollouts/utils/controller.processNextWorkItem(0x1ae6b60, 0xc00000dbc0, 0x183247b, 0x7, 0xc000522e88, 0xc0004c4600, 0xc000000018)
        /go/src/github.com/argoproj/argo-rollouts/utils/controller/controller.go:76 +0x9a
github.com/argoproj/argo-rollouts/utils/controller.RunWorker(0x1ae6b60, 0xc00000dbc0, 0x183247b, 0x7, 0xc000cf7e88, 0xc0004c4600)
        /go/src/github.com/argoproj/argo-rollouts/utils/controller/controller.go:21 +0x5d
github.com/argoproj/argo-rollouts/rollout.(*RolloutController).Run.func1()
        /go/src/github.com/argoproj/argo-rollouts/rollout/controller.go:184 +0x81
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00090b350)
        /go/src/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00090b350, 0x3b9aca00, 0x0, 0xc000000001, 0xc00009a4e0)
        /go/src/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc00090b350, 0x3b9aca00, 0xc00009a4e0)
        /go/src/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/argoproj/argo-rollouts/rollout.(*RolloutController).Run
        /go/src/github.com/argoproj/argo-rollouts/rollout/controller.go:183 +0xae
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xb0 pc=0x14995f7]

goroutine 197 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x105
panic(0x1641b00, 0x299e360)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/argoproj/argo-rollouts/rollout.(*RolloutController).reconcileBlueGreenPause(0xc0000cc300, 0xc0005ff5e8, 0xc0005ff3f0, 0xc000842f00, 0x2)
        /go/src/github.com/argoproj/argo-rollouts/rollout/bluegreen.go:147 +0x87
github.com/argoproj/argo-rollouts/rollout.(*RolloutController).syncReplicasOnly(0xc0000cc300, 0xc0003d0000, 0xc0001af540, 0x2, 0x2, 0x0, 0x0, 0x0)
        /go/src/github.com/argoproj/argo-rollouts/rollout/sync.go:228 +0x678
github.com/argoproj/argo-rollouts/rollout.(*RolloutController).syncHandler(0xc0000cc300, 0xc0006445f0, 0xd, 0x0, 0x0)
        /go/src/github.com/argoproj/argo-rollouts/rollout/controller.go:286 +0x9c9
github.com/argoproj/argo-rollouts/utils/controller.processNextWorkItem.func1(0x1ae6b60, 0xc00000dbc0, 0xc00058de88, 0xc0004c4600, 0x183247b, 0x7, 0x15ce2e0, 0xc00081bd50, 0x0, 0x0)
        /go/src/github.com/argoproj/argo-rollouts/utils/controller/controller.go:65 +0x131
github.com/argoproj/argo-rollouts/utils/controller.processNextWorkItem(0x1ae6b60, 0xc00000dbc0, 0x183247b, 0x7, 0xc000522e88, 0xc0004c4600, 0xc000000018)
        /go/src/github.com/argoproj/argo-rollouts/utils/controller/controller.go:76 +0x9a
github.com/argoproj/argo-rollouts/utils/controller.RunWorker(0x1ae6b60, 0xc00000dbc0, 0x183247b, 0x7, 0xc000cf7e88, 0xc0004c4600)
        /go/src/github.com/argoproj/argo-rollouts/utils/controller/controller.go:21 +0x5d
github.com/argoproj/argo-rollouts/rollout.(*RolloutController).Run.func1()
        /go/src/github.com/argoproj/argo-rollouts/rollout/controller.go:184 +0x81
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00090b350)
        /go/src/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00090b350, 0x3b9aca00, 0x0, 0xc000000001, 0xc00009a4e0)
        /go/src/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc00090b350, 0x3b9aca00, 0xc00009a4e0)
        /go/src/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/argoproj/argo-rollouts/rollout.(*RolloutController).Run
        /go/src/github.com/argoproj/argo-rollouts/rollout/controller.go:183 +0xae
```

**Comment by mathetake** [comment]
I encountered exactly the same problem and found that

- https://github.com/argoproj/argo-rollouts/blob/d32942a163b50c4b9daa122cbbe3ad37464f87de/rollout/sync.go#L216
- https://github.com/argoproj/argo-rollouts/blob/d32942a163b50c4b9daa122cbbe3ad37464f87de/rollout/bluegreen.go#L147

it seems the above two lines are involved with its SEGV.


I am going to work on the patch 👍

**Comment by mathetake** [comment]
i figured out that the following would solve the problem

```
@@ -209,12 +209,13 @@ func (c *RolloutController) syncReplicasOnly(r *v1alpha1.Rollout, rsList []*apps
        logCtx := logutil.WithRollout(r)
        isControllerPaused := len(r.Status.PauseConditions) > 0
        logCtx.Infof("Syncing replicas only (controllerPaused: %v, userPaused %v, isScaling: %v)", isControllerPaused, r.Spec.Paused, isScaling)
-       newRS, oldRSs, err := c.getAllReplicaSetsAndSyncRevision(r, rsList, false)
-       if err != nil {
-               return err
-       }
-       // NOTE: it is possible for newRS to be nil (e.g. when template and replicas changed at same time)
+
        if r.Spec.Strategy.BlueGreen != nil {
+               newRS, oldRSs, err := c.getAllReplicaSetsAndSyncRevision(r, rsList, true)
+               if err != nil {
+                       return err
+               }
+
                roCtx := newBlueGreenCtx(r, newRS, oldRSs)
                previewSvc, activeSvc, err := c.getPreviewAndActiveServices(r)
                if err != nil {
@@ -231,6 +232,11 @@ func (c *RolloutController) syncReplicasOnly(r *v1alpha1.Rollout, rsList []*apps
        // The controller wants to use the rolloutCanary method to reconcile the rolllout if the rollout is not paused.
        // If there are no scaling events, the rollout should only sync its status
        if r.Spec.Strategy.Canary != nil {
+               newRS, oldRSs, err := c.getAllReplicaSetsAndSyncRevision(r, rsList, false)
+               if err != nil {
+                       return err
+               }
+
```

but creating another RS inside of `syncReplicasOnly` function seems strange to me...

lemme find other solutions

**Comment by dthomson25** [comment]
Closed by #314
